### ISSUE TITLE
WkWebview, progress bar, block based kvo warning

### ISSUE THREAD

**Issue by gkmrakesh** [realm/SwiftLint#1989]
### Bug Report
When adding the progress bar for wkwebview like below:

```
extension  WKWebViewController: WKNavigationDelegate {

    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey: Any]?, context: UnsafeMutableRawPointer?) {

        if keyPath == "estimatedProgress" { // listen to changes and updated view
            if webView.estimatedProgress == 1 {
                outActivityView.stopAnimating()
            } else {
                outActivityView.startAnimating()
            }
        }

    }

}
```

##### Warning that swiftlint throwing:

Block Based KVO Violation: Prefer the new block based KVO API with keypaths when using Swift 3.2 or later. (block_based_kvo)

### Environment

* SwiftLint version : 0.24.0
* Installation method used: CocoaPods

Please suggest how to resolve this warning? I have tried adding KVO for "estimatedProgress" but none seems work!

**Comment by ornithocoder** [comment]
Hi @gkmrakesh, this seems a valid violation. Since Swift 3.2 it's possible to use blocks for KVO, instead of overriding `observeValue(forKeyPath:of:change:context:)`.

The new syntax is:

```swift
let observer = foo.observe(\.value, options: [.new, etc....]) { (foo, change) in
    print(change.newValue)
}
```

You can find more about it here: [What's New in Cocoa Touch](https://developer.apple.com/videos/play/wwdc2017/201/) (it starts around 20:00).

For your project you can:

1. adapt the code to use block based KVO, or
2. disable the `block_based_kvo` rule locally on the file, or
3. disable the rule for all the files in your project via `.swiftlint.yml`.

**Comment by gkmrakesh** [comment]
@ornithocoder thank you

I want to fix this warning instead disable it.

```
let observer = webView.observe(\.estimatedProgress, options: .new) { (webview, change) in
            print(change.newValue)
        }
```

I tried multiple ways including one above but none seems to work!!

**Comment by ornithocoder** [comment]
@gkmrakesh what exactly isn't working? SwiftLint still throws a violation or the code doesn't work after the change?

1. If SwiftLint keeps showing the violation after the change, then it's a SwiftLint defect.
1. If the code doesn't work after the change, then it's code related. Use the sample below as example, since observation on `estimatedProgress` should work.

<img width="1382" alt="screen shot 2018-01-04 at 12 57 51 pm" src="https://user-images.githubusercontent.com/19753339/34562550-3c9407b2-f14f-11e7-9b6f-d551342d9847.png">

**Comment by gkmrakesh** [comment]
@ornithocoder thank you

i am facing issue 2 of above post. i have the same code as in your example but it is not working as expected. 

```

import UIKit
import WebKit

class WKWebViewController: UIViewController, WKUIDelegate, WKNavigationDelegate {

    @IBOutlet weak var outHeaderView: UIView!
    @IBOutlet weak var outContainerView: UIView!
    @IBOutlet weak var outActivityView: UIActivityIndicatorView!
    
    typealias DealUrl = HomeState.DealsCardConfig.DealUrl
    
    var webView: WKWebView!
    var availabilityUrl: String?
    
    func newState(state: DealUrl) {
        availabilityUrl = state.url
    }
    
    @IBAction func actGoBack(_ sender: Any) {

        if webView.canGoBack {
            webView.goBack()
        } else {
            appStore.dispatch(ReSwiftRouter.SetRouteAction([Pages.HomeViewControllerPage.rawValue]))
        }
        
    }
        
    private func getWebViewFrame() -> CGRect {
        
        let headerHeight = self.outHeaderView.frame.height
        let webViewHeight = UIScreen.main.bounds.height - headerHeight
        let webViewFrame = CGRect(x: self.view.frame.origin.x, y: headerHeight, width: UIScreen.main.bounds.width, height: webViewHeight)
        
        return webViewFrame
    }
    
    private func initWebView() {
        
        let webConfig = WKWebViewConfiguration()
        let webViewFrame = getWebViewFrame()
        
        webView = WKWebView(frame: webViewFrame, configuration: webConfig)
        webView.uiDelegate = self
        webView.navigationDelegate=self
        
    }
    
    private func addWebViewToContainer() {
        
        outContainerView = webView
        view.addSubview(outContainerView)
        view.addSubview(outActivityView)
        
    }
    
    private func loadAvailURLInWebView() {
        
        if let availUrlStr = availabilityUrl, let availURL = URL(string: availUrlStr) {
            let availRequest = URLRequest(url: availURL)
            webView?.load(availRequest)
        }

    }
    
        private func trackInitialPageLoadProgress() {
            let observer = webView.observe(\.estimatedProgress, options: .new) { (webview, change) in
                **// This is never being called**
                 print(change.newValue)
            }
        }
    
    override func viewDidLoad() {
        super.viewDidLoad()
    
        
        initWebView()
        addWebViewToContainer()
        trackInitialPageLoadProgress()
        loadAvailURLInWebView()
        
    }

}



```


For now I have disabled the swiftlint rule.

**Comment by ornithocoder** [comment]
@gkmrakesh you have to hold a reference to the observer. Add a property to you class to hold it. When it gets out of scope, it automatically stops observing.

Can you close this issue on GitHub since it's not `SwitLint` related? We can keep messaging here even if it's closed. Thanks :-)

**Comment by gkmrakesh** [comment]
Sure thank you

**Comment by ornithocoder** [comment]
@gkmrakesh let me know if holding a reference to the observer fixed the problem. I replicated your class on an iOS Playground and it worked for me.

**Comment by gkmrakesh** [comment]
@ornithocoder This answer worth not less than 2 hours time.

Thank you so much

**Comment by yashwanth777** [comment]
@gkmrakesh @ornithocoder what is holding a reference?? I am facing the same issue
